### ISSUE TITLE
streamline production/pre-release handling

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#665]
The current system, of updating the release notes in Git (or forgetting to) and then copying to NuGet, it tedious and error-prone.
(And the temporary aberration of keeping the "changes in 2.0" in the wiki does not help.)

I think we should do something to reduce the effort.
For example, we could keep the release notes in the source, and they could be updated with PRs, as are the docs.
Then we could see about having the notes automatically copied into the NuGet package when it's packed.
If we couple this with some automated way to attach the NuGet package to the release, we could automatically take a snapshot of the release notes and apply them to the GitHub release at the same time.

A big advantage (to my mind) of making this change is that I think we'd be more likely to update the release notes as PRs are merged, and we'd end up having more eyes on them earlier. Up 'til now, I've been writing most of the release notes, and probably the other owners are taking a peek at them before release, but that's never been explicit in our process.

There are tools to help integrate release notes management into our daily work and then release cycle. Eventually we may want to explore them.

Note that there are some risks with putting the release notes in source, not dissimilar to those of putting the documentation in source.
For example, if a grievous error is found in the release notes, what do we do?

Oh! We could edit the NuGet package and GitHub release notes by hand, and submit a PR to the code. Not great, and the old tagged version of the source would have the wrong info, but that's probably not a huge concern.

Thoughts?

**Comment by adamralph** [comment]
We could also skip the copying of the release notes to the GitHub release and NuGet entirely and just provide a URL to those in source. If we did it in markdown, we'd also get an anchor for that version, and we wouldn't have to worry about when we change them, since the URL could just point to the latest version of that file at any time. Moreover, we wouldn't *necessarily* have to have that file in source. It *could* live in the wiki.

**Comment by blairconrad** [comment]
Everything you say is true.
As a middle ground, if we still wanted to capture the contents of the notes in the GitHub release notes, we could use octokit (I just got that! "kit" is a tiny cat) to push the notes there and change the nuspoec so the release notes are

```
https://github.com/FakeItEasy/FakeItEasy/releases/tag/$Version$
```
And pass the `Version` into the pack command.

**Comment by blairconrad** [comment]
(For that matter, we could use a property to replace the release notes with the actual text, if it were clean enough to read from NuGet. Which it kind of isn't. I like a link better.)

**Comment by adamralph** [comment]
GitHub release notes are even less of an issue than NuGet since they support markdown, so the URL can be a link. NuGet, with it's pitiful plain text only, isn't so nice, but I've seen many other packages do it anyway, e.g. https://www.nuget.org/packages/LibLog/

Even if it's automated, it's still yet another thing (2 things!) to keep in sync.

**Comment by blairconrad** [comment]
Sounds like, based on latest discussion in #666, we want to

1. keep all release notes (analyzer and main package) in the GitHub release
2. have the NuGet Packages' release notes just be https://github.com/FakeItEasy/FakeItEasy/releases

Does that sound right? If so, we can update the nuspec and rakefile.rb and be done with this, I think.

**Comment by adamralph** [comment]
:+1: sounds good to me

**Comment by thomaslevesque** [comment]
> Does that sound right? If so, we can update the nuspec and rakefile.rb and be done with this, I think.

:+1:

**Comment by blairconrad** [comment]
Our decision to just link to GitHub release release notes (yeah, that's what I meant to type) introduces a small inconsistency in the handling of release notes for pre-production builds.
Currently we do not do a full release for the pre-production builds, instead we perform [special steps](https://github.com/FakeItEasy/FakeItEasy/wiki/Managing-releases/bad4d5b6f45fcb96951593ec7c8625757f800c7e#pre-production-beta-releases). As a result, we use a single GitHub release, for the next production build, and continue to accrue release notes in it. Unfortunately, the release notes aren't available to the public.

On reflection, this inconsistency is annoying, and we're not really saving much effort (anything) by having a different procedure, so let's just do a "full release" every time. Here's the "release checklist"  right now:

> - [ ] run code analysis in VS in *Release* mode and address violations (send a regular PR which must be merged before continuing)
> - [ ] **if necessary, change `VERSION_SUFFIX` on [CI Server](http://teamcity.codebetter.com/admin/editBuildParams.html?id=buildType:bt929)
>       to appropriate "-beta123" or "" (for non-betas) value and initiate a build**
> - [ ] check build
> -  edit draft release in [GitHub UI](https://github.com/FakeItEasy/FakeItEasy/releases):
>     - [ ] **complete release notes, mentioning non-owner contributors, if any**
>     - [ ] attach nupkg
>     - [ ] publish the release
> - [ ] push NuGet package
> - [ ] **copy release notes from GitHub to NuGet**
> - [ ] de-list pre-release or superseded buggy NuGet packages if present **(copy any release notes forward to the new version)**
> - [ ] update website with contributors list (if in place)
> - [ ] tweet, mentioning contributors and post link as comment here for easy retweeting ;-)
> - [ ] post tweet in JabbR ([general-chat][1]) and Gitter ([FakeItEasy/FakeItEasy][2])
> - [ ] post links to the NuGet and GitHub release in each issue in this milestone, with thanks to contributors
> - [ ] use `rake set_version[new_version]` to
>     - create a new branch
>     - change CommonAssemblyInfo.cs to expected minor version (of form _xx.yy.zz_)
>     - push to origin
>     - create PR to upstream master
> - [ ] use `rake create_milestone` to
>     - create a new milestone for the next release
>     - create new issue (like this one) for the next release, adding it to the new milestone
>     - create a new draft GitHub Release
> - [ ] close this milestone

If we make the change to do a "full release" every time, we'll need to

1. still only create one milestone - for the production build
2. create a release and "release issue" for all builds as well.
  - One option is to stop using `VERSION_SUFFIX` on the CI server, instead updating the suffix when we run `rake set_version`. This would mean that we'd need to run the step twice or something in order to create e.g. the 3.0 release/issue/milestone and the 3.0.0-beta001 release and issue. And then eventually reset the version back to 3.0.0 before release. Awkward.
  - An alternative would be to introduce a separate command for creating the pre-production paraphernalia, _and also_ continue to update the VERSION_SUFFIX on the CI Server. This is perhaps less terrible, but _slightly_ annoying.
3. move the step to consolidate non-production builds' release notes into the production build's release notes, if applicable up to "edit draft release"
4. remove the step for copying release notes to NuGet. Instead the nuspec will be updated to include a link to the GitHub releases page

I think that's it. Thoughts on the options for step 2?

[1]: https://jabbr.net/#/rooms/general-chat
[2]: https://gitter.im/FakeItEasy/FakeItEasy

**Comment by adamralph** [comment]
> An alternative would be to introduce a separate command for creating the pre-production paraphernalia, and also continue to update the VERSION_SUFFIX on the CI Server. This is perhaps less terrible, but slightly annoying.

I prefer this. I think managing release quality (alpha, beta, RC, RTM) in the source code is more awkward.

**Comment by blairconrad** [comment]
I like it too. Either way it's kind of double work, but I'm hoping we don't pre-release too much. Also, it keeps the commit log cleaner!

**Comment by blairconrad** [comment]
Not "like". "Dislike less".

**Comment by blairconrad** [comment]
@adamralph  and I [chatted with about new rake tasks for managing pre-release versions](https://gitter.im/FakeItEasy?at=57445600eed2f3f916a44f77).

Upshot:

1. combine version-setting with the milestone/release/issue creation for planned production versions. Rake task to be called `next_version`
2. new task `pre_release` to create the pre-release Release and issue, linked to the milestone from step 1

**Comment by blairconrad** [comment]
If/when #739 is merged, I'd update the owner's section of the wiki to talk about how to start a new pre-release release, and amend any in-progress release checklist issues to match as well.

**Comment by blairconrad** [comment]
I've updated [Managing Release](https://github.com/FakeItEasy/FakeItEasy/wiki/Managing-releases). Mostly in the pre-production section, but while I was there, I did change the reference to "rakefile.rb" to a more generic "build task".
I think that's all for this issue. If you'd like to take a gander, I'd be obliged. Otherwise, I can close this after a suitable interval.

**Comment by blairconrad** [comment]
(I forgot to say. I updated #726 as well.)

**Comment by adamralph** [comment]
LGTM, thanks!
